ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -804,6 +810,46 @@ func (r *RayClusterReconciler) updateStatus(instance *rayiov1alpha1.RayCluster)
 	return nil
 }
 
+func (r *RayClusterReconciler) updateDataAnnotation(instance *rayiov1alpha1.RayCluster) error {
+	rayData := make(map[string]string)
+
+	if val, ok := instance.ObjectMeta.Annotations[common.RayDataAnnotationKey]; ok {
+		if err := json.Unmarshal([]byte(val), &rayData); err != nil {
+			r.Log.Info(
+				fmt.Sprintf("unable to parse %s annotation. Overwriting", common.RayDataAnnotationKey),
+				"RayCluster", instance.Name, "namespace", instance.Namespace, "previous data", val,
+			)
+			rayData = make(map[string]string)
+		}
+	}
+
+	if ip, err := r.getHeadNodeIP(instance); err != nil {
+		return fmt.Errorf("failed to get head node ip: %w", err)
+	} else {
+		if ip == "" {
+			r.Log.Info(fmt.Sprintf("head node ip is empty. Not updating %s annotation", common.RayDataAnnotationKey), "RayCluster", instance.Name, "namespace", instance.Namespace)
+		}
+		rayData[common.RayHeadNodeIPAnnotationKey] = ip
+	}
+
+	return nil
+}
+
+func (r *RayClusterReconciler) getHeadNodeIP(instance *rayiov1alpha1.RayCluster) (string, error) {
+	runtimePods := corev1.PodList{}
+	filterLabels := client.MatchingLabels{common.RayClusterLabelKey: instance.Name, common.RayNodeTypeLabelKey: string(rayiov1alpha1.HeadNode)}
+	if err := r.List(context.TODO(), &runtimePods, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+		return "", err
+	}
+	if len(runtimePods.Items) < 1 {
+		return "", fmt.Errorf("unable to find head node pod. cluster name %s, filter labels %v", instance.Name, filterLabels)
+	} else if len(runtimePods.Items) > 1 {
+		return "", fmt.Errorf("found multiple head node pods. cluster name %s, filter labels %v", instance.Name, filterLabels)
+	}
+
+	return runtimePods.Items[0].Status.PodIP, nil
```

Comment:
what if the PodIP is not set?

Reply:
calling function will not update the `status` [here](https://github.com/ray-project/kuberay/pull/468/files#diff-72ecc3ca405f1e828187748d4f1ec8160bccffa2a4f84a364cd7a94a78e1adb9R895-R901).

Reply:
I mean we are expected to return a string, but if the pod exists, and it has no IP assigned yet, then how about we return error.

That way, in the function calling getHeadPodIP, we simply have to check for error, and not worry about an empty value "" .

This is not a blocker, just an observation :)

Reply:
good idea, updated!